### ISSUE TITLE
Building on Arch fails many tests

### ISSUE THREAD

**Issue by cinghiopinghio** [pimutils/todoman#225]
Hi,
I'm experiencing many test failures while updating in archlinux.
They include the the humanized test mentioned [here](https://github.com/pimutils/todoman/issues/216).
I'm trying to install version 3.0.1 from version 2.1.0.

you can see the log [here](https://gist.github.com/cinghiopinghio/28f8f7117e7ceb1f2a29f00616cd5811)

Any clue?

**Comment by hobarrera** [comment]
Yup, tests seem to fail if `python-pyicu` is installed (because one of the test libraries changes it's return values, not because todoman does anything different).

You can skip the tests if you have that installed, but I'll leave an issue open to try to fix this scenario permanently.

**Comment by cinghiopinghio** [comment]
yeah, I remember now I had the same problem with last upgrade, sorry for the noise.

**Comment by cinghiopinghio** [comment]
I removed pyicu and updated, I still have a test failure:

```
================================================================================================================= FAILURES =================================================================================================================
____________________________________________________________________________________________________________ test_default_due2 _____________________________________________________________________________________________________________

tmpdir = local('/tmp/pytest-of-cp/pytest-1/test_default_due20'), runner = <conftest.runner.<locals>.SleepyCliRunner object at 0x7f2c97e05a90>, create = <function create.<locals>.inner at 0x7f2c9591d6a8>
todos = {'aaa': <todoman.model.Todo object at 0x7f2c973f78d0>, 'bbb': <todoman.model.Todo object at 0x7f2c973f7198>, 'ccc': <todoman.model.Todo object at 0x7f2c973f7128>}

    @freeze_time(datetime.datetime.now())
    def test_default_due2(tmpdir, runner, create, todos):
        cfg = tmpdir.join('config')
        cfg.write('default_due = 24\n', 'a')
    
        r = runner.invoke(cli, ['new', '-ldefault', '-dtomorrow', 'aaa'])
        assert not r.exception
        r = runner.invoke(cli, ['new', '-ldefault', 'bbb'])
        assert not r.exception
        r = runner.invoke(cli, ['new', '-ldefault', '-d', 'one hour', 'ccc'])
        assert not r.exception
    
        todos = {t.summary: t for t in todos(all=True)}
        assert todos['aaa'].due.date() == todos['bbb'].due.date()
>       assert todos['ccc'].due == todos['bbb'].due - datetime.timedelta(hours=23)
E       AssertionError: assert FakeDatetime(2017, 3, 31, 18, 27, 10, tzinfo=tzlocal()) == (FakeDatetime(2017, 4, 1, 18, 27, 10, tzinfo=tzlocal()) - datetime.timedelta(0, 82800))
E        +  where FakeDatetime(2017, 3, 31, 18, 27, 10, tzinfo=tzlocal()) = <todoman.model.Todo object at 0x7f2c973f7128>.due
E        +  and   FakeDatetime(2017, 4, 1, 18, 27, 10, tzinfo=tzlocal()) = <todoman.model.Todo object at 0x7f2c973f7198>.due
E        +  and   datetime.timedelta(0, 82800) = <class 'datetime.timedelta'>(hours=23)
E        +    where <class 'datetime.timedelta'> = datetime.timedelta

tests/test_cli.py:277: AssertionError
========================================================================================================== pytest-warning summary ==========================================================================================================
WC1 None pytest_funcarg__httpserver: declaring fixtures using "pytest_funcarg__" prefix is deprecated and scheduled to be removed in pytest 4.0.  Please remove the prefix and use the @pytest.fixture decorator instead.
WC1 None pytest_funcarg__httpsserver: declaring fixtures using "pytest_funcarg__" prefix is deprecated and scheduled to be removed in pytest 4.0.  Please remove the prefix and use the @pytest.fixture decorator instead.
WC1 None pytest_funcarg__smtpserver: declaring fixtures using "pytest_funcarg__" prefix is deprecated and scheduled to be removed in pytest 4.0.  Please remove the prefix and use the @pytest.fixture decorator instead.
=================================================================================== 1 failed, 122 passed, 3 xfailed, 3 pytest-warnings in 10.08 seconds ====================================================================================
==> ERROR: A failure occurred in check().
```

**Comment by hobarrera** [comment]
I'm a bit of at a loss: Out of curiosity, what timezone are you one? I really can't reproduce this (I also use Arch, **and** package the Arch package). It's, honestly the only factory that I can think we might differ on.

**Comment by cinghiopinghio** [comment]
I'm on CEST. I have no clue.

**Comment by cinghiopinghio** [comment]
I tried to navigate the code but this click module is quite obscure to me.
I couldn't find the relevant place where the `-d one hour` is interpreted (this is what the test fails to set up).

**Comment by hobarrera** [comment]
Yeah, click is a bit unintuitive at first. The stack for this call is basically:

[cli.py:277](https://github.com/pimutils/todoman/blob/8b303a900e9e89fe9288222c6fdb0feb1840ba83/todoman/cli.py#L277)
[cli.py:143](https://github.com/pimutils/todoman/blob/8b303a900e9e89fe9288222c6fdb0feb1840ba83/todoman/cli.py#L143)
[cli.py:65](https://github.com/pimutils/todoman/blob/8b303a900e9e89fe9288222c6fdb0feb1840ba83/todoman/cli.py#L65)
[formatters.py:133](https://github.com/pimutils/todoman/blob/8b303a900e9e89fe9288222c6fdb0feb1840ba83/todoman/formatters.py#L133)

Can you try the branch `fix-test_default_due2-test-for-some`? I suspect it *might* fix the failing test, but I'm not entirely sure.

**Comment by cinghiopinghio** [comment]
nope.
I just tried that branch but still I have the same error:

```
E       AssertionError: assert FakeDatetime(2017, 4, 4, 11, 50, 29, tzinfo=tzlocal()) == (FakeDatetime(2017, 4, 5, 11, 50, 29, tzinfo=tzlocal()) - datetime.timedelta(0, 82800))
E        +  where FakeDatetime(2017, 4, 4, 11, 50, 29, tzinfo=tzlocal()) = <todoman.model.Todo object at 0x7ff81e600fd0>.due
E        +  and   FakeDatetime(2017, 4, 5, 11, 50, 29, tzinfo=tzlocal()) = <todoman.model.Todo object at 0x7ff81e6007f0>.due
E        +  and   datetime.timedelta(0, 82800) = <class 'datetime.timedelta'>(hours=23)
E        +    where <class 'datetime.timedelta'> = datetime.timedelta
```

there seems to be an error defining `bbb`, it should be at 10:50 (as I'm on CEST), right?

**Comment by cinghiopinghio** [comment]
We still have errors.
It seems that `ccc` is set to now which is 11:50!
```
>       assert todos['ccc'].due == now + datetime.timedelta(hours=1)
E       AssertionError: assert FakeDatetime(2017, 4, 4, 11, 50, 29, tzinfo=tzlocal()) == (FakeDatetime(2017, 4, 4, 11, 50, 29, tzinfo=tzlocal()) + datetime.timedelta(0, 3600))
E        +  where FakeDatetime(2017, 4, 4, 11, 50, 29, tzinfo=tzlocal()) = <todoman.model.Todo object at 0x7f41d2544be0>.due
E        +  and   datetime.timedelta(0, 3600) = <class 'datetime.timedelta'>(hours=1)
E        +    where <class 'datetime.timedelta'> = datetime.timedelta
```

**Comment by cinghiopinghio** [comment]
I was debugging your code and I found that it seems to be `parsedatetime.Calendar().parse()` fault.
In fact If I run it on the python console with `one hour` as parameter, I get the expected result (now + one hour).
When it run in you code (in the `DefaultFormatter._parse_datetime_naive()` method) I just got the same as `datetime.datetime.now()`. Dunno why...

As a side note, when you `freeze_time` on `datetime.now()` you should add `tzlocal()` for better interpretation of the output. Not important anyway.

**Comment by hobarrera** [comment]
I don't quite understand. Can you provide a sample script of the failure?

Also, what's you're output for `env`. I wonder if something locale-related might be involved (TBH, it's the only think I can think of that would make our setups differ).

**Comment by cinghiopinghio** [comment]
I found the problem: `freezegun` doesn't work very well with timezones, `parsedatetime.Calendar().parse('one hour')` give different results inside or outside a `freezer`.
There seems to be a PR https://github.com/spulec/freezegun/pull/136 (not sure if it solve our problem) but it's awaiting for merging since more then one year...

I'm not sure if the fault is due to the summer time (which push the clock one hour forward) or the time zone (which push the clock one more hour forward) since time inside the `freezer` is one hour earlier.

**Comment by hobarrera** [comment]
Thanks for tracking that down.

It looks like we can't do much to work around this -- I'll try to skip those tests if DST is active on the machine running them for the next release. I'll also make sure all tests dates are outside DST.

**Comment by cinghiopinghio** [comment]
Bom dia @hobarrera,
just to warn you: I just tried upgrading from `3.0.1` to `3.2` and I've found the same test failure.
No updates on https://github.com/spulec/freezegun side.

**Comment by hobarrera** [comment]
Can you try running `export TZ=UTC` before running tests? That might be a quick hack to fixing the PKGBUILD for now.

**Comment by cinghiopinghio** [comment]
Yes!!!
It does work with this workaround! great!

**Comment by dvzrv** [comment]
@hobarrera here's my makepkg dump. I'm still unable to pass check().
I never use the TZ variable anywhere, but my timezone is set system-wide as by the setup standards of Arch Linux.
Your PKGBUILD however sets it right before using the tests, so in that environment TZ=UTC is set.

[todoman.err.txt](https://github.com/pimutils/todoman/files/1013618/todoman.err.txt)

**Comment by hobarrera** [comment]
@dvzrv That's #119: tests fail if python-pyicu is installed. I believe a **test** dependency of ours has a bug if that's installed. I haven't gotten around to it yet.

I'll close this issue since the problem it described has been fixed, and we have #119 to track that other one anyway (I do suggest you follow up on that one).
